src/rust/engine/serverset/src/lib.rs

```diff
@@ -0,0 +1,456 @@
+      initial_lame_millis,
+      ratio,
+      max_lame_millis,
+    })
+  }
+}
+
+impl<T: Clone + Send + Sync + 'static> Serverset<T> {
+  pub fn new(servers: Vec<T>, backoff_config: BackoffConfig) -> Result<Self, String> {
+    if servers.is_empty() {
+      return Err("Must supply some servers".to_owned());
+    }
+
+    Ok(Serverset {
+      inner: Arc::new(Inner {
+        servers: servers
+          .into_iter()
+          .map(|s| Backend {
+            server: s,
+            unhealthy_info: Arc::new(Mutex::new(None)),
+          }).collect(),
+        next: AtomicUsize::new(0),
+        backoff_config,
+      }),
+    })
+  }
+
+  ///
+  /// Get the next (probably) healthy backend to use.
+  ///
+  /// The caller will be given a backend to use, and should call ServerSet::callback with the
+  /// supplied token, and the observed health of that backend.
+  ///
+  /// If the callback is not called, the health status of the server will not be changed from its
+  /// last known status.
+  ///
+  /// If all resources are unhealthy, the returned Future will delay until a resource becomes
+  /// healthy.
+  ///
+  /// No efforts are currently made to avoid a thundering heard at few healthy servers (or the the
+  /// first server to become healthy after all are unhealthy).
+  ///
+  pub fn next(&self) -> BoxFuture<(T, CallbackToken), String> {
+    let now = Instant::now();
+    let server_count = self.inner.servers.len();
+
+    let mut earliest_future = None;
+    for _ in 0..server_count {
+      let i = self.inner.next.fetch_add(1, Ordering::Relaxed) % server_count;
+      let server = &self.inner.servers[i];
+      let unhealthy_info = server.unhealthy_info.lock();
+      if let Some(ref unhealthy_info) = *unhealthy_info {
+        // Server is unhealthy. Note when it will become healthy.
+
+        let healthy_at = unhealthy_info.healthy_at();
+
+        if healthy_at > now {
+          let healthy_sooner_than_previous = if let Some((_, previous_healthy_at)) = earliest_future
+          {
+            previous_healthy_at > healthy_at
+          } else {
+            true
+          };
+
+          if healthy_sooner_than_previous {
+            earliest_future = Some((i, healthy_at));
+          }
+          continue;
+        }
+      }
+      // A healthy server! Use it!
+      return futures::future::ok((server.server.clone(), CallbackToken { index: i })).to_boxed();
+    }
+    // Unwrap is safe because if we hadn't populated earliest_future, we would already have returned.
+    let (index, instant) = earliest_future.unwrap();
+    let server = self.inner.servers[index].server.clone();
+    // Note that Delay::new_at(time in the past) gets immediately scheduled.
+    Delay::new_at(instant)
+      .map_err(|err| format!("Error delaying for serverset: {}", err))
+      .map(move |()| (server, CallbackToken { index }))
+      .to_boxed()
+  }
+
+  pub fn callback(&self, callback_token: CallbackToken, health: Health) {
```

Comment:
nit: Now that these are actually exposed, giving them purposeful names would be good... `report_health` and `HealthReportToken` or something maybe?

Reply:
Oh, could also mark the token "#[must_use]"